File: sol/2349/2349.cpp
#include <unordered_map>

struct NumberContainers {
    unordered_map<int, int> m_map;
    unordered_map<int, vector<int>> m_indexHeap;

    NumberContainers() {}
    
    void change(int index, int number) {
        m_map[index] = number;
        m_indexHeap[number].push_back(-index);
        ranges::push_heap(m_indexHeap[number]);
    }
    
    int find(int number) {
        if (!m_indexHeap.contains(number)){
            return -1;
        }
        while (!m_indexHeap[number].empty()){
            int idx = -m_indexHeap[number][0];
            if (m_map[idx] == number){
                return idx;
            }
            ranges::pop_heap(m_indexHeap[number]);
            m_indexHeap[number].pop_back();
        }
        return -1;
    }
};

File: sol/2349/2349.py
import heapq as hq
from collections import defaultdict

class NumberContainers:
    def __init__(self):
        self.indexDict: dict[int, int] = dict()  # index -> number
        self.indexHeap: defaultdict[int, list[int]] = defaultdict(list)  # number -> heap[index]

    def change(self, index: int, number: int) -> None:
        self.indexDict[index] = number
        hq.heappush(self.indexHeap[number], index)

    def find(self, number: int) -> int:
        while len(self.indexHeap[number]) != 0:
            idx = self.indexHeap[number][0]
            if self.indexDict[idx] == number:
                return idx
            hq.heappop(self.indexHeap[number])
        return -1

